File: milbot/plugins/help.py
import re
import slack

from log import Log
import utils


@slack.RTMClient.run_on(event="message")
async def help_func(**payload):
    """コマンド一覧をつくる"""

    data = payload["data"]
    if utils.is_bot_message(data):
        return
    web_client = payload["web_client"]
    channel_id = data.get("channel")
    text = data.get("text")

    try:
        if re.match(r"^milbot help hidden", text, re.IGNORECASE):
            await web_client.chat_postMessage(
                channel=channel_id,
                text=mes_hidden()
            )
        elif re.match(r"^milbot help", text, re.IGNORECASE):
            await web_client.chat_postMessage(
                channel=channel_id,
                text=mes()
            )
    except Exception as e:
        Log.error(e)
        raise(e)


def mes():
    return """以下のコマンドを受け付けます(｀･ω･´)
`milbot help`
`milbot help hidden`
`milbot ping`
`milbot peng`
`milbot peng help`
`milbot atnd`
`milbot atnd help`
`milbot atnd add`
`milbot atnd delete`
`milbot atnd list`
`帰宅の木`
`milbot kitakunoki help`

また以下の語句に反応します(｀･ω･´)
`帰宅の木`
`帰宅の木の苗`
575
57577
7775
"""


def mes_hidden():
    return """隠しコマンドです。絶対に実行しないでください(´･ω･｀)
`milbot exit`
`milbot exit help`
`milbot bash`
`milbot bash help`
`milbot python3`
`milbot python3 help`
"""

File: milbot/plugins/help/help.go
package help

import (
	"context"
	"regexp"

	"github.com/high-moctane/milbot/milbot/botutils"
	"github.com/nlopes/slack"
)

// help を発動する先頭文字列
var helpPrefix = regexp.MustCompile(`(?i)^milbot help`)

var helpMessage = "以下のコマンドを受け付けています。\n" +
	"    `milbot atnd`\n" +
	"    `milbot atnd add`\n" +
	"    `milbot atnd delete`\n" +
	"    `milbot atnd help`\n" +
	"    `milbot atnd list`\n" +
	"    `milbot exit`\n" +
	"    `milbot exit help`\n" +
	"    `milbot help`\n" +
	"    `milbot kitakunoki help`\n" +
	"    `milbot peng`\n" +
	"    `milbot peng help`\n" +
	"    `milbot ping`\n" +
	"    `milbot ping help`\n" +
	"    `milbot restart`\n" +
	"    `milbot restart help`\n" +
	"    `milbot script help`\n" +
	"    `milbot bash`\n" +
	"    `milbot python3`\n" +
	"    `milbot verse help`\n" +
	"\n" +
	"また以下の機能があります。\n" +
	"    帰宅の木\n" +
	"    575 警察\n"

// Plugin の中身は必要ない
type Plugin struct{}

// New でプラグインを作成する
func New() Plugin {
	return struct{}{}
}

// Serve では "milbot help" に反応して help を返す
func (p Plugin) Serve(ctx context.Context, api *slack.Client, ch <-chan slack.RTMEvent) {
	for {
		select {
		case <-ctx.Done():
			return

		case msg := <-ch:
			switch ev := msg.Data.(type) {
			case *slack.MessageEvent:
				// bot かどうかを判定
				if ev.BotID != "" {
					continue
				}

				if helpPrefix.MatchString(ev.Text) {
					go help(api, ev)
				}
			}
		}
	}
}

func help(api *slack.Client, ev *slack.MessageEvent) {
	botutils.SendMessageWithLog(api, ev.Channel, helpMessage)
}
